src/txn/txn_db.rs

```diff
@@ -0,0 +1,729 @@
+// Copyright 2021 Yiyuan Liu
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use std::{
+    collections::BTreeMap,
+    ffi::CString,
+    fs, iter,
+    marker::PhantomData,
+    path::{Path, PathBuf},
+    ptr,
+    sync::{Arc, RwLock},
+};
+
+use crate::{
+    column_family::UnboundColumnFamily, db::DBAccess, db_options::OptionsMustOutliveDB, ffi,
+    ffi_util::to_cpath, AsColumnFamilyRef, BoundColumnFamily, ColumnFamilyDescriptor,
+    DBIteratorWithThreadMode, DBRawIteratorWithThreadMode, Direction, Error, IteratorMode, Options,
+    ReadOptions, SnapshotWithThreadMode, Txn, TxnDBOptions, TxnOptions, WriteBatch, WriteOptions,
+    DB, DEFAULT_COLUMN_FAMILY_NAME,
+};
+use libc::{c_char, c_int, size_t};
+
+/// RocksDB TransactionDB.
+///
+/// Please read the official [guide](https://github.com/facebook/rocksdb/wiki/Transactions)
+/// to learn more about RocksDB TransactionDB.
+pub struct TxnDB {
+    pub(crate) inner: *mut ffi::rocksdb_transactiondb_t,
+    cfs: RwLock<BTreeMap<String, Arc<UnboundColumnFamily>>>,
```

Comment:
i know this is a bit of stretch for you, but could you try to follow the `SingleThreaded` and `MultiThreaded` mechanism for transaction db as well? so that `cfs` doesn't necessarily need to be wrapped inside `RwLock`?

Reply:
I don't really understand why thread mode is needed, is it for performance reasons?

If we follow the mechanism in `DBWithThreadMode`, we will have typename such as `TransactionDBWithThreadMode<MultiThreaded>` and `OptimisticTransactionDBWithThreadMode<MultiThreaded>`. That looks ugly.

Reply:
Not necessarily, the naming convention could be changed - I think the type should have still been called `DB` but should have been given a default generic parameter to prevent ugliness.

Reply:
> Not necessarily, the naming convention could be changed - I think the type should have still been called `DB` but should have been given a default generic parameter to prevent ugliness.

This has done in commit 5c7e0d9.